### PR TITLE
icons on custom legends

### PR DESCRIPTION
**Pull request by matallo**
close https://github.com/CartoDB/cartodb/issues/10726

**steps to test this:**

- create a map, add a custom legend, and pick an icon in any of the items
- check the icon (and the color) changes both in Builder and the map

screen shot 2017-01-10 at 14 59 41

this needs to be reviewed and merged first:

- [x] cartodb.js support https://github.com/CartoDB/cartodb.js/pull/1545

### CODE REVIEW

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/legend/category-item.js [hunk 1]
```diff
@@ -1,3 +1,47 @@
 var Backbone = require('backbone');
 
-Backbone.Form.editors.List.CategoryModel = Backbone.Form.editors.NestedModel;
+Backbone.Form.editors.List.CategoryModel = Backbone.Form.editors.NestedModel.extend({
```
Comment (nobuti) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/legend/category-item.js]: Could you explain a bit the logic here? I'm not an expert in Backbone.Form, at all, I need to understand it. The construction seems a bit weird, rendering a form inside an form editor.

Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/legend/category-item.js]: yeah, it was hard for me to create a mental model of this too, but it is a nested form, I'm just extending it in order to have access and pass `usermodel`, `configmodel`, and `modals` options

the original model comes from backbone-forms https://github.com/powmedia/backbone-forms/blob/master/src/editors/nestedmodel.js

Comment (nobuti) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/legend/category-item.js]: I feel this is too complicated. Why not passing those params as model options as we do in https://github.com/CartoDB/cartodb/blob/master/lib/assets/core/javascripts/cartodb3/editor/widgets/widgets-form/widgets-form-factory.js#L48-L55? This options should go in the form model we are using https://github.com/CartoDB/cartodb/blob/master/lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/legend/form/legend-nested-category-definition-form-model.js

Comment (nobuti) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/legend/category-item.js]: After a deeper look, you are absolutely right, there is no way to pass options to the nested model, and this is the way to go.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/legend/category-item.js [hunk 2]
```diff
@@ -1,3 +1,47 @@
...
+
+  initialize: function (options) {
+    Backbone.Form.editors.Base.prototype.initialize.call(this, options);
+    Backbone.Form.editors.Base.prototype._setOptions.call(this, options); // Options
+
+    if (!this.form) throw new Error('Missing required option "form"');
+    if (!options.schema.model) throw new Error('Missing required "schema.model" option for NestedModel editor');
+
+    this._userModel = this.options.userModel;
```
Comment (nobuti) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/legend/category-item.js]: I think is too specific, I'm thinking in a case when we need other/more dependencies.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/legend/category-item.js [hunk 3]
```diff
@@ -1,3 +1,47 @@
...
+
+  render: function () {
+    // Get the constructor for creating the nested form; i.e. the same constructor as used by the parent form
+    var NestedForm = this.form.constructor;
+
+    var data = this.value || {};
+    var NestedModel = this.schema.model;
+
+    // Wrap the data in a model if it isn't already a model instance
+    var modelInstance = (data.constructor === NestedModel) ? data : new NestedModel(data, {
```
Comment (nobuti) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/legend/category-item.js]: Related to comment above, I'd go with something like this, if possible:
```
var modelInstance = (data.constructor === NestedModel) ? data : new NestedModel(data, this.options));
```

**Comment by nobuti** [comment]
Let's fix the conflicts.

**Comment by matallo** [comment]
YAY

**Comment by nobuti** [comment]
Frontend test passing 🚀

**Pull request merged**